### PR TITLE
Cleanup IoBuf and add source/sink for it

### PR DESCRIPTION
**Pull request by dranikpg**
Some things that I desperately missed in DF:
- IoBuf is not movable (And is wrapped often into unique_ptr to account for this 🤯 )
- IoBuf WriteAndCommit akin to ReadAndConsume
- IoBuf cannot be used as a sink or as a source
- There is no way to convert Bytes into string_view easily

### CODE REVIEW

File: base/io_buf.h [hunk 1]
```diff
@@ -44,85 +56,61 @@ class IoBuf {
...
-      }
-      delete[] buf_;
-    }
-
-    buf_ = nb;
-    capacity_ = sz;
-  }
+  // Reserve by whole buffer size.
+  // Use EnsureCapacity instead for resizing only by desired append buffer size.
+  void Reserve(size_t full_size);
```
Comment (dranikpg) [inline base/io_buf.h:91]: Those function is dangerously misleading, it accounts only for the TOTAL size. Most of the time EnsureCapacity is really needed from outside accessors

File: io/io.cc [hunk 2]
```diff
@@ -73,6 +73,22 @@ void AsyncWriteState::OnCb(Result<size_t> res) {
   owner->AsyncWriteSome(cur, arr.end() - cur, [this](Result<size_t> res) { this->OnCb(res); });
 }
 
+// Read some bytes from buf starting at given offset offs and update it.
+Result<size_t> ReadSomeBytes(const iovec* v, uint32_t len, Bytes buf, off_t* offs) {
```
Comment (romange) [inline io/io.cc:77]: please rename v into dest, it's hard to understand from the signature what's the input and what's the output.

File: base/io_buf.h [hunk 3]
```diff
@@ -44,85 +56,61 @@ class IoBuf {
...
+    size_ += num_written;
   }
 
-  void Clear() {
-    size_ = 0;
-    offs_ = 0;
-  }
+  // Copy num_copy bytes from source to append buffer and mark them as written.
+  // Ensures append buffer is large enough.
+  void WriteAndCommit(size_t num_copy, const void* source);
```
Comment (romange) [inline base/io_buf.h:82]: usually a ptr is followed by length

File: base/io_buf.h (hunk 1 above)
Comment (romange) [reply base/io_buf.h:91]: if you think it's not needed, feel free to remove it...

**Review by romange — APPROVED** [review]

**Pull request merged**